### ISSUE TITLE
Multipart uploads trigger notification too early

### ISSUE THREAD

**Issue by GeoffreyBooth** [localstack/localstack#501]
I have some code that calls the AWS SDK `s3.upload` ([API](http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#upload-property), [blog post](https://aws.amazon.com/blogs/developer/announcing-the-amazon-s3-managed-uploader-in-the-aws-sdk-for-javascript/)), which automatically splits large files into parts to upload to S3 in parallel. On the bucket that receives this upload, I have a notification set to invoke a lambda function. When that lambda function runs, the first thing it tries to do is use `s3.getObject` to open the uploaded file. LocalStack returns an error, saying that the object cannot be found.

I looked through the logs, and from these I can deduce what’s happening:

```bash
127.0.0.1 - - [08/Dec/2017 19:20:20] "PUT /inbox/2017-12-08/upload.zip?partNumber=2
&uploadId=WvwbL3Z%2FmHNTV0q87lp4AuWUQ6b8WAa9hEKCqJHZFuUeOAtBSNoLqTgg HTTP/1.1" 200 -

127.0.0.1 - - [08/Dec/2017 19:20:20] "PUT /inbox/2017-12-08/upload.zip?partNumber=1
&uploadId=WvwbL3Z%2FmHNTV0q87lp4AuWUQ6b8WAa9hEKCqJHZFuUeOAtBSNoLqTgg HTTP/1.1" 200 -

2017-12-08T19:20:20:DEBUG:localstack.services.awslambda.lambda_executors: Running lambda cmd: 
CONTAINER_ID="$(docker create  -e STATIC_ROOT="$STATIC_ROOT" -e LANDING_ROOT="$LANDING_ROOT" 
-e MEDIA_ROOT="$MEDIA_ROOT" -e HOSTNAME="$HOSTNAME" -e SMTP_HOST="$SMTP_HOST" 
-e APP_ROOT="$APP_ROOT" -e AWS_LAMBDA_EVENT_BODY="$AWS_LAMBDA_EVENT_BODY" 
-e LOCALSTACK_HOSTNAME="$LOCALSTACK_HOSTNAME" "lambci/lambda:nodejs6.10" "index.uploadHandler")";
docker cp "/tmp/localstack/zipfile.0a4e0882/." "$CONTAINER_ID:/var/task";
docker start -a "$CONTAINER_ID";

2017-12-08T19:20:20:DEBUG:localstack.services.awslambda.lambda_executors: Running lambda cmd: 
CONTAINER_ID="$(docker create  -e STATIC_ROOT="$STATIC_ROOT" -e LANDING_ROOT="$LANDING_ROOT" 
-e MEDIA_ROOT="$MEDIA_ROOT" -e HOSTNAME="$HOSTNAME" -e SMTP_HOST="$SMTP_HOST" 
-e APP_ROOT="$APP_ROOT" -e AWS_LAMBDA_EVENT_BODY="$AWS_LAMBDA_EVENT_BODY" 
-e LOCALSTACK_HOSTNAME="$LOCALSTACK_HOSTNAME" "lambci/lambda:nodejs6.10" "index.uploadHandler")";
docker cp "/tmp/localstack/zipfile.0a4e0882/." "$CONTAINER_ID:/var/task";
docker start -a "$CONTAINER_ID";

127.0.0.1 - - [08/Dec/2017 19:20:24] "GET /inbox/2017-12-08/upload.zip HTTP/1.1" 404 -

127.0.0.1 - - [08/Dec/2017 19:20:24] "GET /inbox/2017-12-08/upload.zip HTTP/1.1" 404 -
```

Basically, my 6MB file is uploaded in two parts in parallel; and the lambda function is invoked twice, once for each part. Unfortunately LocalStack has yet to finish merging the parts when either lambda function tries to open the file, so LocalStack returns 404 to both invocations.

A workaround for now is to use `s3.putObject` in place of `s3.upload`; `putObject` does not automatically parallelize uploads, and the notifications behave as expected. But there are two bugs here that should be fixed:
- Multipart uploads should trigger notifications only once at the end, not after each part is uploaded.
- A notification after a successful multipart upload should happen after the parts have been merged and the object is available for use.

**Comment by whummer** [comment]
Thanks for the detailed bug report @GeoffreyBooth , great analysis. Given that you have already analyzed the issue in detail, would you be able to give it a try and raise a PR with the proposed changes? I'd be happy to help out if needed. Thanks!

**Comment by whummer** [comment]
Should be fixed in #1001 - please re-open this issue if the problem persists. Thanks
